ansible_base/resource_registry/tasks/sync.py

```diff
@@ -0,0 +1,411 @@
+from __future__ import annotations  # support python<3.10
+
+import asyncio
+import csv
+import time
+from dataclasses import dataclass, field
+from datetime import timedelta
+from enum import Enum
+from io import StringIO, TextIOBase
+from typing import TypedDict
+
+from asgiref.sync import sync_to_async
+from django.conf import settings
+from django.db import transaction
+from django.db.models import QuerySet
+from django.db.utils import DatabaseError, IntegrityError
+from django.utils import timezone
+from requests import HTTPError
+
+from ansible_base.resource_registry.models import Resource, ResourceType
+from ansible_base.resource_registry.rest_client import ResourceAPIClient
+
+
+class ManifestNotFound(HTTPError):
+    """Raise when server returns 404 for a manifest"""
+
+
+class ResourceDeletionError(DatabaseError):
+    """Raise for deletion errors on Django ORM"""
+
+
+class ResourceSyncHTTPError(HTTPError):
+    """Custom catchall error"""
+
+
+# Type Aliases
+ResourcePK = AnsibleID = ResourceHash = ServiceID = ModifiedTime = str
+
+
+class ResourceTypeName(str, Enum):
+    # keep the order
+    ORGANIZATION = "shared.organization"
+    TEAM = "shared.team"
+    USER = "shared.user"
+
+    def __str__(self):
+        return self.value
+
+
+class SyncStatus(str, Enum):
+    CREATED = "created"
+    UPDATED = "updated"
+    NOOP = "noop"
+    CONFLICT = "conflict"
+    UNAVAILABLE = "unavailable"
+
+    def __str__(self):
+        return self.value
+
+
+@dataclass
+class ManifestItem:
+    ansible_id: AnsibleID
+    resource_hash: ResourceHash
+    modified: ModifiedTime
+    service_id: ServiceID | None = None
+    resource_data: dict | None = None
+    # additional_data: dict | None = None
+
+    def __hash__(self):
+        return hash(self.ansible_id)
+
+
+@dataclass
+class SyncResult:
+    __match_args__ = ('status', 'item')
+
+    status: SyncStatus
+    item: ManifestItem
+
+    def __iter__(self):
+        """Allows unpacking  status, item = SyncResult(...)"""
+        return iter((self.status, self.item))
+
+
+class APIClientConfig(TypedDict):
+    service_url: str
+    service_path: str
+    verify_https: bool
+    requests_auth_kwargs: dict
+    raise_if_bad_request: bool
+
+
+# Functions
+
+
+def create_api_client(config: APIClientConfig | None = None) -> ResourceAPIClient:
+    """Factory for pre-configured ResourceAPIClient."""
+    config = config or settings.RESOURCE_PROVIDER_CONFIG
+    config["raise_if_bad_request"] = False
+    return ResourceAPIClient(**config)
```

Comment:
This actually depends on #311 beign merged so the token based rest_client can be used from that PR

Reply:
For now, if one wants to run locally, on the service settings.py add this

```python
RESOURCE_PROVIDER_CONFIG = {
    "service_url": "https://localhost:8000",
    "service_path": "/api/gateway/v1/service-index/",
    "verify_https": False,
    "requests_auth_kwargs": {"auth": ("admin", "admin")},
    "raise_if_bad_request": True,
}

```